File: Python/Easy/trees/min_depth_binary_tree.py
"""

Given a binary tree, find its minimum depth.

The minimum depth is the number of nodes along the shortest path from the 
root node down to the nearest leaf node.

Note: A leaf is a node with no children.

"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root: return 0
        left = self.minDepth(root.left)
        right = self.minDepth(root.right)
        if root.right is None and root.left is None:
            return 1
        if root.right is None:
            return 1 + left
        if root.left is None:
            return 1 + right
        return min(left,right) + 1

File: JavaScript/Easy/trees/min_depth_binary_tree.js
/**
 * 
 * @param {TreeNode} root
 * @return {number}
 * 
 */

const TreeNode = (val, left, right) => {
    this.val = (val === undefined ? 0 : val)
    this.left (left === undefined ? null : left)
    this.right (right === undefined ? null : right)
}

var minDepth = (root) => {
    if (root == null) {
        return 0
    } else {
        let left = minDepth (root.left)
        let right = minDepth (root.right)
        if (root.right == null && root.left == null){
            return 1
        }
        if (root.right == null){
            return 1 + left
        }
        if (root.left == null) {
            return 1 + right
        }
        return Math.min(left,right) + 1
    }
}